File: 백준/Gold/15683. 감시/감시.py
from re import L
import sys
input = sys.stdin.readline
import collections
import copy
n, m = map(int,input().split())

maze = [list(map(int,input().split())) for _ in range(n)]

dx = [0,-1,0,1]
dy = [1,0,-1,0]

stack = collections.deque([maze])
min_count = float('inf')
while stack:

  this_turn_maze = stack.popleft()
  
  flag = False
  for i in range(n):
    for j in range(m):
      if this_turn_maze[i][j] == 1:
        flag = True
        for k in range(4):
          new_maze = copy.deepcopy(this_turn_maze)
          new_maze[i][j] = 7
          idx = 1
          row = i
          col = j
          while 0 <= row + dx[k]*idx < n and 0 <= col + dy[k]*idx < m:
            # 벽
            if new_maze[row+dx[k]*idx][col+dy[k]*idx] == 6:
              break
            elif 1<= new_maze[row+dx[k]*idx][col+dy[k]*idx]<=5:
              idx +=1
              
            else:
              new_maze[row+dx[k]*idx][col+dy[k]*idx] = 7
              idx += 1
          stack.append(new_maze)  
        break
      elif this_turn_maze[i][j] == 2:
        flag = True
        for t in range(2):
          new_maze = copy.deepcopy(this_turn_maze)
          new_maze[i][j] = 7
          for k in range(2):
            idx = 1
            row = i
            col = j
            while 0 <= row + dx[k*2+t]*idx < n and 0 <= col + dy[k*2+t]*idx < m:
              # 벽
              if new_maze[row+dx[k*2+t]*idx][col+dy[k*2+t]*idx] == 6:
                break
              elif 1<= new_maze[row+dx[k*2+t]*idx][col+dy[k*2+t]*idx]<=5:
                idx +=1
              else:
                new_maze[row+dx[k*2+t]*idx][col+dy[k*2+t]*idx] = 7
                idx += 1
          stack.append(new_maze)  
  
        break
      elif this_turn_maze[i][j] == 3:
        flag = True
        for t in range(4):
          new_maze = copy.deepcopy(this_turn_maze)
          new_maze[i][j] = 7
          
          for k in range(2):
            idx = 1
            row = i
            col = j
            while 0 <= row + dx[(t+k)%4]*idx < n and 0 <= col + dy[(t+k)%4]*idx < m:
              # 벽
              if new_maze[row+dx[(t+k)%4]*idx][col+dy[(t+k)%4]*idx] == 6:
                break
              elif 1<= new_maze[row+dx[(t+k)%4]*idx][col+dy[(t+k)%4]*idx]<=5:
                idx +=1
              else:
                new_maze[row+dx[(t+k)%4]*idx][col+dy[(t+k)%4]*idx] = 7
                idx += 1
          stack.append(new_maze)  
          
        break
      elif this_turn_maze[i][j] == 4:
        flag = True
        for t in range(4):
          new_maze = copy.deepcopy(this_turn_maze)
          new_maze[i][j] = 7
          
          for k in range(3):
            idx = 1
            row = i
            col = j
            while 0 <= row + dx[(t+k)%4]*idx < n and 0 <= col + dy[(t+k)%4]*idx < m:
              # 벽
              if new_maze[row+dx[(t+k)%4]*idx][col+dy[(t+k)%4]*idx] == 6:
                break
              elif 1<= new_maze[row+dx[(t+k)%4]*idx][col+dy[(t+k)%4]*idx]<=5:
                idx +=1
              else:
                new_maze[row+dx[(t+k)%4]*idx][col+dy[(t+k)%4]*idx] = 7
                idx += 1
          stack.append(new_maze)  
          
        break
      elif this_turn_maze[i][j] == 5:
        flag = True
       
        new_maze = copy.deepcopy(this_turn_maze)
        new_maze[i][j] = 7
        t = 0
        for k in range(4):
          idx = 1
          row = i
          col = j
          while 0 <= row + dx[(t+k)%4]*idx < n and 0 <= col + dy[(t+k)%4]*idx < m:
            # 벽
            if new_maze[row+dx[(t+k)%4]*idx][col+dy[(t+k)%4]*idx] == 6:
              break
            elif 1<= new_maze[row+dx[(t+k)%4]*idx][col+dy[(t+k)%4]*idx]<=5:
              idx +=1
            else:
              new_maze[row+dx[(t+k)%4]*idx][col+dy[(t+k)%4]*idx] = 7
              idx += 1
          stack.append(new_maze)  
          
        break
    if flag:
      break
  # 이놈은 이제 시시티비가 없네

  if not flag:
    
    count = 0
    for i in range(n):
      for j in range(m):
        
        if this_turn_maze[i][j] == 0:
          count += 1
      
    min_count = min(count, min_count)
  
print(min_count)

File: 백준/Gold/15683. 감시/감시.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static int n,m;
    static int[][] map;

    static int[] dx = {-1, 0, 1, 0};
    static int[] dy = {0, 1, 0, -1};
//    static Set<Integer>[][]visited;
    static int[] cctv;

    static List<int[]> cctvLoc;
    static int[] proc;
    static int cnt = Integer.MAX_VALUE;

    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        map = new int[n][m];
//        visited = new HashSet[n][m];
        cctvLoc = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
                if (1 <= map[i][j] && map[i][j] <= 5) {
//                    visited[i][j] = new HashSet<>();
                    cctvLoc.add(new int[]{i, j});
                }
            }
        }
        proc = new int[cctvLoc.size()];

        cctv = new int[6];
        cctv[1] = 2;
        cctv[2] = 10;
        cctv[3] = 3;
        cctv[4] = 11;
        cctv[5] = 15;
        count(0);
        System.out.println(cnt);

    }

    private static void count(int index) {
//        System.out.println(index);
        if (index == cctvLoc.size()) {
//            System.out.println(Arrays.toString(proc));
            int result = 0;
            int[][] copy = new int[n][m];
            for (int r = 0; r < n; r++) {
                copy[r] = map[r].clone();
            }
            for (int i = 0; i < cctvLoc.size(); i++) {
                int thisRow = cctvLoc.get(i)[0], thisCol = cctvLoc.get(i)[1];
                int bit = proc[i];
//                System.out.println(thisRow + " " + thisCol+" "+proc[i]);
                for (int k = 0; k < 4; k++) {
                    if ((bit & (1 << k)) != 0) {
                        int idx = 1;
                        while (true) {
                            int row = thisRow + dx[k] * idx;
                            int col = thisCol + dy[k] * idx;
//                            System.out.println(row + " " + col);
                            if (0 <= row && row < n && 0 <= col && col < m) {
                                if (copy[row][col] == 6) {
                                    break;
                                } else if (copy[row][col] == 0) {
                                    copy[row][col] = 7;
                                    idx++;
                                } else {
                                    idx ++;
                                }
                            } else {
                                break;
                            }
                        }
                    }
                }


            }
            for (int row = 0; row < n; row++) {
                for (int col = 0; col < m; col++) {
                    result += copy[row][col] == 0 ? 1 : 0;
                }
            }
//            System.out.println(result);
            cnt = Math.min(result, cnt);

            return;
        }
        int[] thisCctvLoc = cctvLoc.get(index);
        int thisRow = thisCctvLoc[0];
        int thisCol = thisCctvLoc[1];
        int cctvSort = cctv[map[thisRow][thisCol]];
        Set<Integer> visited = new HashSet<>();

        while (true) {
            if (!visited.contains(cctvSort)) {
                proc[index] = cctvSort;
                visited.add(cctvSort);
                count(index + 1);
                cctvSort = shiftLeft(cctvSort);
            } else {
                break;
            }
        }

    }



    private static int shiftLeft(int num) {
        int value = num;
        num = (num << 1) & 15;
        if ((value << 1) > 15) {
            num |= 1;
        }
         return num;


    }
}